ddd-core/src/main/kotlin/io/paju/ddd/AggregateRoot.kt

```diff
@@ -25,11 +32,39 @@ constructor(val id: AggregateRootId)
         applyChange(event, true)
     }
 
-    protected fun applyChange(event: E, isNew: Boolean) {
-        apply(event)
-
+    private fun applyChange(event: E, isNew: Boolean) {
+        state = apply(event, state)
         if (isNew) {
             changes.add(event)
         }
     }
-}
\ No newline at end of file
+
+    class Builder<out A: AggregateRoot<S, E>, S: State, E : StateChangeEvent>
+    internal constructor (private val constructor: () -> A)
+    {
+        fun newInstance(): A {
+            return constructor()
+        }
+
+        fun newInstanceWithCreateEvent(createEvent: E): A {
+            return constructor().apply {
+                applyChange(createEvent, true)
+            }
+        }
```

Comment:
I'm thinking naming of this. "CreateEvent" implies what event would somehow contribute to the creation of the new aggregate, but I think this is more generic - it's new instance and apply combined, the event doesn't have any special meaning other than it's the first event. Creation related steps have to be part of the constructor passed as function because otherwise newInstance() would render invalid aggregate. Maybe rename this to newInstanceAndApplyEvent() so that's maybe more clear that this creates new instance and then applies *any* of the aggregate's events?

Reply:
More more thing - actually this event has special meaning in a sense that developer has to make sure the event passed on is something that can be applied to a new instance. Some events can be applied successfully only in certain states. For replay this shouldn't be a problem, but here it might be. This makes me thinking do we need this? Aggregate constructor can already customize initialization logic  and in normal use you would get new instance and then call public business function to change state and aggregate would enforce business rules.  

If we'd like to replay aggregate completely from events, then I guess we need own create aggregate event which contains UUID and own event stream structure that guarantees that create event is always first in the stream for replaying. 

I think we can leave this as I think we can live with the fact that to replay, you'd need UUID first and then replay the events.

Reply:
In short because of these concerns, I'd drop newInstanceWithCreateEvent()

Reply:
This is something that i have wondered how it should be done or even should it not be done. 

Question for is that:

1. Do we need special `Notification`-event in the event stream that tells that the aggregate instance is created?
2. Or is it just the first event (could be any event) that tells that the aggregate instance is created?

Now i think that if we need `1.` logic then it should be implemented some other way. I will remove this because it is not ready.

Comment about this:

> If we'd like to replay aggregate completely from events, then I guess we need own create aggregate event which contains UUID and own event stream structure that guarantees that create event is always first in the stream for replaying.

This is what i tried to do with this implementation. Current implemention is not good, because it is up to developer when to use `create event`. 

I thing that aggregate UUID should be in envelope and not in events in event stream.